pkg/controller/deployment/util/deployment_util.go

```diff
@@ -495,19 +495,24 @@ func getReplicaSetFraction(logger klog.Logger, rs apps.ReplicaSet, d apps.Deploy
 		return -*(rs.Spec.Replicas)
 	}
 
-	deploymentReplicas := *(d.Spec.Replicas) + MaxSurge(d)
-	annotatedReplicas, ok := getMaxReplicasAnnotation(logger, &rs)
-	if !ok {
+	deploymentMaxReplicas := *(d.Spec.Replicas) + MaxSurge(d)
+	deploymentMaxReplicasBeforeScale, ok := getMaxReplicasAnnotation(logger, &rs)
+	if !ok || deploymentMaxReplicasBeforeScale == 0 {
```

Comment:
This place (out of 2) is the only change in this function. The rest are renames.

Reply:
This extra guard lgtm to make sure we don't divide by 0, so it improves code readability, but I'm wondering if this is also a fix. Is this scenario (getting 0 here possible e2e)?

Reply:
If this turns out to be a bugfix, I would prefer to separate the fix and cleanup (renames) PRs.

Reply:
It is a fix, but a mild one.

For the context:


The ReplicaSet can have the annotation set to 0 only if the Deployment is set to 0 and  as the RS is also 0 so we would never get to this point in the normal flow:

https://github.com/kubernetes/kubernetes/blob/6ae2d359ad2eed549f467016465d04fa4b222b7a/pkg/controller/deployment/util/deployment_util.go#L471

The impact of the bug:

1. Fortunately, dvision by 0 in floats results in Infinity and no panic

https://github.com/kubernetes/kubernetes/blob/6ae2d359ad2eed549f467016465d04fa4b222b7a/pkg/controller/deployment/util/deployment_util.go#L510

2. We convert it to max int here:

https://github.com/kubernetes/kubernetes/blob/6ae2d359ad2eed549f467016465d04fa4b222b7a/pkg/controller/deployment/util/deployment_util.go#L511

3. But we keep it in the bounds for the possible scale here:

https://github.com/kubernetes/kubernetes/blob/6ae2d359ad2eed549f467016465d04fa4b222b7a/pkg/controller/deployment/util/deployment_util.go#L478-L487

4. The bug results in incorrect fractional scaling:


So for this to become an issue the user would have to tamper with the annotation (remove it or change it) and the deployment status is not synced (e.g. 10 pods failed). The impact would then just be an incorrect fractional scaling. E.g.

rs1: 50 -> 67
rs2: 50 (40+10) -> 53

instead of the correct answer

rs1: 50 -> 60
rs2: 50 (40+10) -> 60


Since it requires user intervention during the fractional scaling and failing pods/unsynced deployment status, I think it should be enough to just unit test it.

The impact is, editing the annotation will mess up your scaling. Which can also happen in other instances after this fix.

Also, I think the renames help to document the issue better.

Reply:
Thank you for the investigation and the detailed summary. In that case I'm ok to bundle the renames and the fix together. It probably does not hurt to still annotate the PR with "/kind bug" and also split the two aspects into separate commits. WDYT?

Reply:
Ok. I have split the commits and updated the description.